app/model/SubscriptionAcquisitionComponents.scala

```diff
@@ -0,0 +1,181 @@
+package model
+
+import com.gu.acquisition.model.{OphanIds, ReferrerAcquisitionData}
+import com.gu.acquisition.typeclasses.AcquisitionSubmissionBuilder
+import com.gu.memsub.Benefit.{PaperDay, Weekly}
+import com.gu.memsub.BillingPeriod
+import com.gu.memsub.BillingPeriod.{Month, Quarter, SixMonths, Year}
+import com.gu.memsub.promo.{PercentDiscount, Promotion}
+import com.gu.stripe.Stripe.Charge
+import com.typesafe.scalalogging.LazyLogging
+import forms.SubscriptionsForm
+import ophan.thrift.event.PaymentProvider.{Gocardless, Stripe}
+import ophan.thrift.event._
+import play.api.libs.json._
+
+import scala.collection.Set
+
+//import com.gu.acquisition.model.ReferrerAcquisitionData.referrerAcquisitionDataReads
+
+case class SubscriptionAcquisitionComponents(
+  subscribeRequest: SubscribeRequest,
+  promotion: Option[Promotion.AnyPromotion],
+  acquisitionDataJSON: Option[String]
+)
+
+
+object SubscriptionAcquisitionComponents {
+
+  implicit object subscriptionAcquisitionSubmissionBuilder
+    extends AcquisitionSubmissionBuilder[SubscriptionAcquisitionComponents] with LazyLogging {
+
+    def buildOphanIds(components: SubscriptionAcquisitionComponents): Either[String, OphanIds] = {
+      import components._
+
+      Right(
+        OphanIds(
+          pageviewId = subscribeRequest.genericData.ophanPageViewId,
+          visitId = subscribeRequest.genericData.ophanVisitId,
+          browserId = subscribeRequest.genericData.ophanBrowserId
+        )
+      )
+    }
+
+    private def parseJsonSafely(json: String): Option[JsValue] = {
+      try {
+        Some(Json.parse(json))
+      } catch {
+        case scala.util.control.NonFatal(t) =>
+          logger.warn(s"""Unable to parse "$json" as JSON. ${t.getMessage()}""")
+          None
+      }
+    }
+
+    private def referrerAcquisitionDataFromJSON(json: String): Option[ReferrerAcquisitionData] = {
```

Comment:
Had a play around - here is what a Scalaz implementation could look like...

```scala
private def referrerAcquisitionDataFromJSON(json: String): Option[ReferrerAcquisitionData] = {
  import \/._
  fromTryCatchNonFatal(Json.parse(json))
    .leftMap(err => logger.warn(s"""Unable to parse "$json" as JSON. ${err.getMessage}"""))
    .flatMap { jsValue =>
      Json.fromJson[ReferrerAcquisitionData](jsValue).fold(
        err => left(logger.warn(s"Unable to decode JSON $jsValue to an instance of ReferrerAcquisitionData. ${JsError.toJson(err)}")),
        data => right(data)
      )
    }
    .toOption
}
```
Not sure if it's preferable?

Reply:
yeah, I reckon it is! 💎